### PR TITLE
[SPRAL] add recipe

### PR DESCRIPTION
**Pull request by amontoison**
superseded #2130
@ViralBShah I updated your script, let's see if it works for all platforms :crossed_fingers:

### CODE REVIEW

**Comment by ViralBShah** [comment]
I think you need to use METIS4 and not METIS (which is 5).

**Comment by amontoison** [comment]
SPRAL has an interface for METiS4 (https://github.com/lanl-ansi/spral/blob/master/src/metis4_wrapper.F90) and METIS5 (https://github.com/lanl-ansi/spral/blob/master/src/metis5_wrapper.F90).

**Comment by ViralBShah** [comment]
Maybe see if METIS 4 interface has better luck?

**Comment by amontoison** [comment]
I have the same failing platforms with METIS4.

**Comment by amontoison** [comment]
C_LONG is Int32 on 32-bits systems, they never built on it...

**Comment by amontoison** [comment]
@ViralBShah all tests passed :tada: 
Do you know how to do a dynamic version of `libspral.a` ?

**Comment by giordano** [comment]
Something like
```
cc -shared -fPIC -o "${libdir}/libspral.${dlext}" -Wl,$(flagon --whole-archive) "${prefix}/lib/libspral.a"
```

**Comment by dpo** [comment]
This is what I used in ASL: https://github.com/JuliaPackaging/Yggdrasil/blob/master/A/ASL/build_tarballs.jl#L22

**Comment by giordano** [comment]
[`flagon`](https://github.com/JuliaPackaging/Yggdrasil/blob/bb679f74dd1ce39bd1bf8a2669bbdbfffb179f2c/0_RootFS/Rootfs/bundled/utils/flagon) is there exactly to avoid doing that.

**Comment by dpo** [comment]
I see. That's great. I wasn't aware there was a specific shortcut for that kind of operation! Thanks!

**Comment by giordano** [comment]
I think I created the script after seeing the hoops you were jumping through :stuck_out_tongue:

File: S/SPRAL/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="METIS_jll", uuid="d00139f3-1899-568f-a2f0-47f597d42d70"))
+    Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae"))
+    Dependency(PackageSpec(name="OpenBLAS32_jll", uuid="656ef2d0-ae68-5445-9ca0-591084a874a2"))
+    Dependency(PackageSpec(name="Hwloc_jll", uuid="e33a78d0-f292-5ffc-b300-72abe9b543c8"))
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version = v"8.1.0")
```
Comment (giordano) [inline S/SPRAL/build_tarballs.jl:60]: Would you be OK to require Julia v1.6?  It'll make it easier to add new platforms in the future.
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version = v"8.1.0", julia_compat="1.6")
```

**Comment by amontoison** [comment]
@giordano I tried 
```
cc -shared -fPIC -o "${libdir}/libspral.${dlext}" -Wl,$(flagon --whole-archive) "${prefix}/lib/libspral.a"
```
 but it's not working, can you have a look at it ?

**Comment by giordano** [comment]
What's the error?

**Comment by amontoison** [comment]
> What's the error?

I saved the output of the command :
[error.txt](https://github.com/JuliaPackaging/Yggdrasil/files/6410609/error.txt)

**Comment by giordano** [comment]
This library is a mess

**Comment by giordano** [comment]
My understanding is that they don't believe in shared libraries at all.  How important is to get a shared library?  You'd like to call into libspral or it's only needed as build dependency for other packages?  In the former cases you should really raise the issue upstream.

**Comment by dpo** [comment]
Yes, we would like to call into SPRAL. I'll contact them.

**Comment by dpo** [comment]
@amontoison Would you mind opening a PR to update SPRAL with updated configure scripts, etc., as fixed by `update_configure_scripts` and (if applicable) your patches? I can't even get past the `configure` step on my Mac and I'm not able to build your recipe locally either.

**Comment by amontoison** [comment]
We met the team that works on SPRAL this morning with @dpo. We updated the main repository to include the patches and fix some issues with recent gcc compilers.

We are working together to build shared libraries.

**Comment by dpo** [comment]
This works for me on macOS:
```diff
diff --git a/S/SPRAL/build_tarballs.jl b/S/SPRAL/build_tarballs.jl
index 26ec8944..f9942e75 100644
--- a/S/SPRAL/build_tarballs.jl
+++ b/S/SPRAL/build_tarballs.jl
@@ -24,7 +24,10 @@ CFLAGS=-fPIC CPPFLAGS=-fPIC CXXFLAGS=-fPIC FFLAGS=-fPIC FCFLAGS=-fPIC \
     ../configure --prefix=${prefix} --build=${MACHTYPE} --host=${target} \
     --with-blas="-L${libdir} -lopenblas" --with-lapack="-L${libdir} -lopenblas" \
     --with-metis="-L${libdir} -lmetis" --with-metis-inc-dir="${prefix}/include"
-make && make install
+make
+gfortran -fPIC -shared -Wl,-all_load libspral.a -Wl,-noall_load -lgomp -lopenblas -lhwloc -lmetis -lstdc++ -o libspral.${dlext}
+make install
+cp libspral.${dlext} ${libdir}
 exit
 """
 
@@ -32,7 +35,6 @@ exit
 # platforms are passed in on the command line
 platforms = expand_gfortran_versions(supported_platforms())
 
-
 # The products that we will ensure are always built
 products = [
     ExecutableProduct("spral_ssids", :spral_ssids)
```

**Comment by amontoison** [comment]
@giordano
Can I remove the last `-n` here : https://github.com/JuliaPackaging/Yggdrasil/blob/bb679f74dd1ce39bd1bf8a2669bbdbfffb179f2c/0_RootFS/Rootfs/bundled/utils/flagon#L76 ?

@dpo It works fine !

**Comment by amontoison** [comment]
> [`flagon`](https://github.com/JuliaPackaging/Yggdrasil/blob/bb679f74dd1ce39bd1bf8a2669bbdbfffb179f2c/0_RootFS/Rootfs/bundled/utils/flagon) is there exactly to avoid doing that.

... when It works :wink:

File: S/SPRAL/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+cd build
+CFLAGS=-fPIC CPPFLAGS=-fPIC CXXFLAGS=-fPIC FFLAGS=-fPIC FCFLAGS=-fPIC \
+    ../configure --prefix=${prefix} --build=${MACHTYPE} --host=${target} \
+    --with-blas="-L${libdir} -lopenblas" --with-lapack="-L${libdir} -lopenblas" \
+    --with-metis="-L${libdir} -lmetis" --with-metis-inc-dir="${prefix}/include"
+make
+gfortran -fPIC -shared -Wl,$(flagon --whole-archive) libspral.a -Wl,$(flagon --no-whole-archive) -lgomp -lopenblas -lhwloc -lmetis -lstdc++ -o libspral.${dlext}
+make install
+cp libspral.${dlext} ${libdir}
+exit
```
Comment (giordano) [inline S/SPRAL/build_tarballs.jl:31]: https://github.com/JuliaPackaging/Yggdrasil/blob/8be19820afa24ff817185f3c53c4d20786a8add3/S/SHTOOLS/build_tarballs.jl#L28

```suggestion
gfortran -fPIC -shared -Wl,$(flagon --whole-archive) libspral.a -Wl,$(flagon --no-whole-archive | cut -d' ' -f1) -lgomp -lopenblas -lhwloc -lmetis -lstdc++ -o ${libdir}/libspral.${dlext}
make install
```

File: 0_RootFS/Rootfs/bundled/utils/flagon [hunk 3]
```diff
@@ -73,7 +73,7 @@ for flag in "${FLAGS[@]}"; do
         --no-whole-archive)
             case "${PLATFORM}" in
                 *-apple-*)
-                    echo -n "-noall_load -n"
+                    echo -n "-noall_load "
```
Comment (giordano) [inline 0_RootFS/Rootfs/bundled/utils/flagon:76]: If I remember correctly, the correct thing to do is to not print anything

Comment (giordano) [reply 0_RootFS/Rootfs/bundled/utils/flagon]: #2988

**Pull request merged**